io.openems.edge.common/src/io/openems/edge/common/powercharacteristic/AbstractPowerCharacteristic.java

```diff
@@ -0,0 +1,195 @@
+	 * @param alias           the Component Alias
+	 * @param enabled         is the Component enabled?
+	 * @param meterId         the Meter-ID
+	 * @param noOfBufferHours the number of buffer hours to make sure the battery
+	 *                        still charges full, even on prediction errors
+	 */
+	protected void activate(ComponentContext context, String id, String alias, boolean enabled) {
+		super.activate(context, id, alias, enabled);
+	}
+
+	/**
+	 * Initialize the P by U characteristics.
+	 *
+	 * <p>
+	 * Parsing JSON then putting the point variables into pByUCharacteristicEquation
+	 *
+	 * <pre>
+	 * [
+	 *  { "voltageRatio": 0.9, 		"power" : -4000 }},
+	 *  { "voltageRatio": 0.93,		"power": -1000 }},
+	 *  { "voltageRatio": 1.07,		"power": 0}},
+	 *  { "voltageRatio": 1.1, 		"power": 1000 } }
+	 * ]
+	 * </pre>
+	 * 
+	 * @param percentQ the configured Percent-by-Q values
+	 * @return
+	 * 
+	 * @throws OpenemsNamedException on error
+	 */
+
+	private Map<Float, Float> initialize(String powerConfig) throws OpenemsNamedException {
+		Map<Float, Float> voltagePowerMap = new HashMap<>();
+		try {
+			JsonArray jpowerV = JsonUtils.getAsJsonArray(JsonUtils.parse(powerConfig));
+			for (JsonElement element : jpowerV) {
+				Float powerConfValue = (float) JsonUtils.getAsInt(element, "power");
+				float voltageRatioConfValue = JsonUtils.getAsFloat(element, "voltageRatio");
+				voltagePowerMap.put(voltageRatioConfValue, powerConfValue);
+			}
+		} catch (NullPointerException e) {
+			throw new OpenemsException("Unable to set values [" + powerConfig + "] " + e.getMessage());
+		}
+		return voltagePowerMap;
+	}
+
+	public Integer getPowerLine(String powerConfig, float ratio) throws OpenemsNamedException {
+
+		Map<Float, Float> voltagePowerMap = this.initialize(powerConfig);
+		float linePowerValue = this.getValueOfLine(voltagePowerMap, ratio);
+
+		return (int) linePowerValue;
+	}
+
+	private float getValueOfLine(Map<Float, Float> qCharacteristic, float gridVoltageRatio) {
+		float m = 0, t = 0;
+		// find to place of grid voltage ratio
+		Comparator<Entry<Float, Float>> valueComparator = (e1, e2) -> e1.getKey().compareTo(e2.getKey());
+		Map<Float, Float> map = qCharacteristic.entrySet().stream().sorted(valueComparator)
+				.collect(Collectors.toMap(Entry::getKey, Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+		List<Float> voltageList = new ArrayList<Float>(map.keySet());
+		List<Float> powerList = new ArrayList<Float>(map.values());
+		// if the grid voltage ratio in the list, return that point
+		for (int i = 0; i < voltageList.size(); i++) {
+			if (voltageList.get(i) == gridVoltageRatio) {
+				int power = (int) powerList.get(i).intValue();
+				return power;
+			}
+		}
+		Point smaller = getSmallerPoint(qCharacteristic, gridVoltageRatio);
+		Point greater = getGreaterPoint(qCharacteristic, gridVoltageRatio);
+		if (greater.x == smaller.x && greater.y == smaller.y) {
+			return smaller.y;
+		}
```

Comment:
If I'm not wrong all of the following code can be replaced with this - by smartly using a TreeMap:
```
	private float getValueOfLine(TreeMap<Float, Float> qCharacteristic, float gridVoltageRatio) {
		Entry<Float, Float> floorEntry = qCharacteristic.floorEntry(gridVoltageRatio);
		if(floorEntry.getKey().equals(gridVoltageRatio)) {
			return floorEntry.getValue().intValue();
		}
		Entry<Float, Float> ceilingEntry = qCharacteristic.ceilingEntry(gridVoltageRatio);
		float m = (float) ((ceilingEntry.getValue() - floorEntry.getValue()) / (ceilingEntry.getKey() - floorEntry.getKey()));
		float t = (float) (floorEntry.getKey() - m * floorEntry.getKey());
		return m * gridVoltageRatio + t;
	}
```
In any case: please add a JUnit test for this.